### PR TITLE
Emit defineProperty calls before param prop assignments

### LINKED ISSUE #34942: Instance properties declaration and initialization in constructor failed with useDefineForClassFields turned on
**TypeScript Version:**  3.7.2

**Search Terms:** useDefineForClassFields, constructor

**Code**

```ts
class Test {
    constructor(public readonly something: number) { }
}

const k = new Test(5);
```

**Expected behavior:** I expect `k.something` to be `5`

**Actual behavior:** `k.something` is undefined, cause TS:
1. writes the value to the property (as it was before)
2. and then rewrite it (with `writable: true`, I think it is bug too) to `void 0`

```js
class Test {
    constructor(something) {
        this.something = something;
        Object.defineProperty(this, "something", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
    }
}
```

**Playground Link:** https://www.typescriptlang.org/play/?useDefineForClassFields=true&target=7&ssl=3&ssc=2&pln=1&pc=1#code/MYGwhgzhAEAqCmEAu0DeAoaXrAPYDtkAnAV2CVyIAoAHEgIxAEthoj4wATAkAT2gi4AtvCQALJvgDmALmj4SQ+vCIBKNNAC+6TUA

### PR DESCRIPTION
**Pull request by sandersn**
Note that I restricted this to when --useDefineForClassFields is true. Nothing changes when it's off. I think this is the correct fix for a patch release.

However, in principal there's nothing wrong with moving parameter property initialisation after property declaration initialisation. It would be Extremely Bad and Wrong to rely on this working:

```ts
class C {
  p = this.q // q is already initialised here!
  constructor(public q: number) { }
}
```

But today it emits the below code, and probably somebody relies on it without knowing:

```js
class C {
    constructor(q) {
        this.q = q;
        this.p = p;
    }
}
```

Fixes #34942

### CODE REVIEW

**Comment by ajafff** [comment]
> probably somebody relies on it

👋 a lot of code in my current project relies on this behavior.
IIRC the order is documented somewhere so changing it would be a huge breaking change.

**Comment by DanielRosenwasser** [comment]
Please do not screw with initialization order of parameter properties vs. property initializers. We're planning on putting this in 3.7.3.

**Comment by sandersn** [comment]
@ajafff could it be in the spec? I'll have to take a look. If that's the case the long-term fix would need to be an interleaving of parameter property defineProperty/assignments, all happening before the defineProperty calls for other property declarations.

**Comment by sandersn** [comment]
Or, after discussing with @RyanCavanaugh, removing the assignments and using the parameter as the initial `value` of the defineProperty call.

**Comment by Jessidhia** [comment]
That'd work as a quick fix for now, but I'd like to raise an issue with the semantics of parameter properties to begin with. The way it's specified in https://github.com/microsoft/TypeScript/blob/master/doc/spec.md#831-constructor-parameters is insufficiently clear, and the most direct interpretation of it actually means it should behave like I wrote in the first paragraph -- in your first `class C` snippet, `p = this.q` should actually raise a "use before initialization" error with `--strictNullChecks`.

I wrote this in a bug report that got closed as a duplicate:

> This depends on how the parameter property is going to be interpreted. If it's still going to be interpreted as the same as writing an assignment in the constructor's body, then it needs to be reordered to happen _after_ all of the `Object.defineProperty` initializers (as they technically run before `super()` returns, or before the constructor block begins evaluating for non-subclassing classes), and `bar`'s initializer should detect `this.foo` as used before initialization.
>
> If they're going to be "as powerful as" actual class fields, then constructor parameter properties must be hoisted higher than any actual class field initializers. Even with this latter interpretation, which is _almost_ what happened, the `this.foo = foo` assignment is incorrect and should be done either immediately after the `Object.defineProperty` for `foo` runs, or the `Object.defineProperty` should use the constructor parameter as `value` initializer.

But for the spec issue itself and how to handle it going forward, should I open a new issue?

**Comment by sandersn** [comment]
@Jessidhia I think the right thing is to have `Object.defineProperty` use the constructor parameter as the `value` initializer. I *think* this resolves the conflict with ecma262 initialisation order.

Although this is complex enough that perhaps we should have a new issue to find out if we need changes to accomodate the class fields proposal.

**Comment by sandersn** [comment]
@rbuckton mind taking a look at this? This is the fix we talked about yesterday

@Jessidhia We decided that we won't make changes to parameter properties unless the TC39 committee considers a similar feature. The feature as specified is kind of weird and fits badly with the ECMAScript spec, but changing it will certainly break people. I don't think it's worthwhile to improve it unless we're forced to.

**Comment by ajafff** [comment]
> We decided that we won't make changes to parameter properties unless the TC39 committee considers a similar feature. The feature as specified is kind of weird and fits badly with the ECMAScript spec, but changing it will certainly break people. I don't think it's worthwhile to improve it unless we're forced to.

Does that mean you always need to use `Object.defineProperty` for classes with parameter properties even with `useDefineForClassFields: true, target: esnext`? As far as I understand the spec, TypeScript is not able to place the parameter property initialisation before initialisation of (native) class fields.

**Comment by DanielRosenwasser** [comment]
🔔 reviewers can we get some eyes on this so we can get 3.7.3 out sooner?

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-3.7

File: tests/cases/conformance/classes/propertyMemberDeclarations/definePropertyESNext.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// @target: esnext
```
Comment (rbuckton) [inline tests/cases/conformance/classes/propertyMemberDeclarations/definePropertyESNext.ts]: You can just merge this with `definePropertyES5.ts` (and rename it?) and use this:
```ts
// @target: es5, esnext
```

**Comment by sandersn** [comment]
@typescript-bot cherry-pick this into release-3.7

**Pull request merged**